db-engine/src/main/java/com/splicemachine/db/impl/sql/conn/GenericLanguageConnectionContext.java

```diff
@@ -4148,13 +4148,22 @@ private String getLogHeader() {
     private static final String SQLHASH_STRING = "sqlHash=";
     private static final String STATEMENT_STRING = ", statement=[ ";
 
-    private String formatLogStmt(String statement) {
+    private String formatLogStmt(String statement, Activation activation) {
         if (statement == null) {
             return "sqlHash=null, statement=null";
         }
-        synchronized (this) {
+        boolean hasBaseActivation = (activation instanceof BaseActivation);
+        Object synchronizationObject = hasBaseActivation ? activation : this;
+        synchronized (synchronizationObject) {
             // cache formatted statement log
-            if (statement.equals(lastLogStmt)) {
+            BaseActivation baseActivation = null;
+            if (hasBaseActivation) {
+                baseActivation = (BaseActivation) activation;
+                if (statement.equals(baseActivation.getLastLogStmt())) {
+                    return baseActivation.getLastLogStmtFormat();
+                }
+            }
+            else if (statement.equals(lastLogStmt)) {
```

Comment:
I think it would be cleaner to use a small guava Cache that handles concurrency for us. If performance wise it is equivalent I'd choose that.

Reply:
Good point.  They probably do better than using "synchronized".  Added caches in BaseActivation and GenericLanguageConnectionContext which are lazily allocated.